README.md
# Quantum Cat-sweeper

This is an almost regular Minesweeper that runs on the [IBM Quantum computer](https://quantumexperience.ng.bluemix.net/qx/experience) or simulator. 

# Rules of the game:
- Click with the left mouse button to discover tiles
- The number on tiles indicates how many bombs is around the tile you clicked
- Don't explode the cats (bombs), they have 50% chance of exploding (may the quantum randomness be in you favor !)
- To win the game, find the golden kitty (who moves based on whether you reveal a tile that isn't a bomb !)

![Front Screen](https://github.com/desireevl/quantum-catsweeper/blob/master/images/mainscreen.PNG)

## Game Over
![Game Over](https://github.com/alexandrebis/quantum-catsweeper/blob/feature/flag-and-tiles/images/lost.png)

## Game Won
![Game Over](https://github.com/alexandrebis/quantum-catsweeper/blob/feature/flag-and-tiles/images/won.PNG)

# Technical Explanation

## Cat Bombs
A Hadamard gate is used on the qubit that represents the bomb and when the tile is clicked, the qubit is measured. It has a 50/50 chance of evaluating to a 1 (bomb explodes) or a 0 (bomb defuses).

## Golden Cat
The golden cat moves around based on your click. If you find it, you win 100% of the time. 

If you click on a tile that is not an exploding bomb nor a blank tile (number tile, defused cat) then the golden cat moves one space in the direction of the tile you just clicked.

If a negatively evaluated bomb is clicked, the golden cat moves one space away.

This knowledge can be used as strategy to win the game. 

# install and launching Guide

Note : The current Qiskit dependency version is a really old version = 0.5.7. To update the version, you must modify the existing code and use up-to-date equivalents. For example, QuantumProgram is deprecated, you must now separately create a circuit and launch the `execute(circuit, backend)` method. [Source](https://qiskit.org/documentation/release_notes.html)

Python 3 is needed

(Optional but strongly recommended) set up a venv with the command :
```bash
python -m venv venv
source venv/Scripts/activate
```

Install the dependencies with :
```bash
pip install -r requirements.txt
``` 

As a temporary workaround, you have to modify your qiskit dependency manually.

Go to : `\venv\lib\site-packages\qiskit\dagcircuit\` and replace all occurences of `.node[` with `._node[` in the following files :
- _dagcircuit.py
- _dagunroller.py

Launch with :
```bash
python main.py
python main.py debug # For debugging mode
```


CHANGELOG
# Changelog

## Last version (30/11/2020)

### Added

- Minesweeper logic
- Number tiles
- Put a flag on a tile (right mouse click)

### Removed

- Reveal groups
- Negative eval for tiles

## First version

### Changes

- Lowered the sound and disabled the loop of the main music
- Changed `LEFT_KEY_BUTTON` to `LEFT_KEY`, then to `LEFT_MOUSE_BUTTON` because why use the left arrow?
- Replace quantum random by regular random and remove it from requirements
- Set `pyxel.mouse(True)` to see the cursor
- Remove IBMQ dependencies as it is not used
- Set last working numpy version
- Update README

### Fixes
- Commented the `qiskit.register()` method as it is deprecated. Will update Qiskit in following versions
- Changed the order of initialization of window to display the game
- Move `images` location for the game to load


### Changes ideas

Use one of the IBM Quantum backends
Use a more recent version of Qiskit
Implement real minesweeper logic
Correctly size the window

qcatsweeper/quantum_logic.py
import random
from enum import Enum
from qcatsweeper import qconfig
from qiskit import *


class TileItems(Enum):
    BLANKS = 0

    TILE1 = 1
    TILE2 = 2
    TILE3 = 3
    TILE4 = 4
    TILE5 = 5
    TILE6 = 6
    TILE7 = 7
    TILE8 = 8
    TILE_NUMBERS = [TILE1, TILE2, TILE3, TILE4, TILE5, TILE6, TILE7, TILE8]

    BOMB_UNEXPLODED = 9
    BOMB_EXPLODED = 10

    GOLDEN_CAT = 11
    FLAG = 12

    POS_EVAL = 42
    NEG_EVAL = -42

    BOMB_DEFUSED = 84


real_device = True
shots = 1024

if real_device:
    try:
        IBMQ.save_account(qconfig.APItoken, overwrite=True)
        IBMQ.load_account()
        provider = IBMQ.get_provider(hub='ibm-q')
        device = provider.get_backend('ibmq_qasm_simulator')
    except qiskit.providers.ibmq.api.exceptions.RequestsApiError:
        print("Could not load IBMQ account. Using local qasm_simulator.\nError : ", sys.exc_info()[1])
        device = Aer.get_backend('qasm_simulator')
else:
    device = Aer.get_backend('qasm_simulator')


# qiskit.register(qconfig.APItoken, qconfig.config["url"])


def get_one_or_zero(grid_script, q, c):

    # measuring qubit and finding which value has the most outcomes
    grid_script.measure(q, c)
    results = execute(grid_script, device, shots=shots).result()
    results_counts = results.get_counts(grid_script)
    print("results_counts", results_counts)

    if len(results_counts) == 1:
        print("return 0")
        return 0
    if results_counts['0'] > results_counts['1']:
        print("return 0")
        return 0
    print("return 1")
    return 1


def new_game_grid(grid_size, bomb_no=20):
    game_grid = [[TileItems.BLANKS for i in range(grid_size)] for j in range(grid_size)]

    # classical random number generator for debugging
    bomb_xy = [random.randint(0, grid_size-1) for i in range(bomb_no * 2)]
    bomb_xy = [bomb_xy[i:i + 2] for i in range(0, bomb_no * 2, 2)]
    # add Bombs
    for coord in bomb_xy:
        if len(coord) > 0:
            game_grid[coord[0]][coord[1]] = TileItems.BOMB_UNEXPLODED

    # add number Tiles
    game_grid = add_number_tiles(game_grid, grid_size)

    # add golden Cat
    game_grid[random.randint(0, grid_size - 1)][random.randint(0, grid_size - 1)] = TileItems.GOLDEN_CAT

    return game_grid


def add_number_tiles(game_grid: list, grid_size: int):
    for row in range(len(game_grid)):
        for col in range(len(game_grid[row])):
            if game_grid[row][col] != TileItems.BOMB_UNEXPLODED and game_grid[row][col] != TileItems.GOLDEN_CAT:
                number = number_of_bombs(row, col, game_grid, grid_size) # Compte les bombes autour (le "gros" du code)
                if number > 0:
                    game_grid[row][col] = TileItems['TILE' + str(number)]

    return game_grid


def number_of_bombs(row, col, game_grid, grid_size: int):
    # if  row == 0 and col == 0: #coin haut et gauche
    bombs = 0

    # On parcourt toutes les cases autours
    for y in range(row - 1, row + 2):
        for x in range(col - 1, col + 2):
            if (y >= 0) and (y < grid_size) and (x >= 0) and (x < grid_size): # Si la case existe
                if game_grid[y][x] == TileItems.BOMB_UNEXPLODED: # Si la case est une bombe
                    bombs += 1
    return bombs


def onclick(clicked_tile):
    """
    params:
    clicked_tile: tile type of the clicked tile
    num_click: number of times a group has been clicked
    """
    q = QuantumRegister(1)
    c = ClassicalRegister(1)
    gridScript =  QuantumCircuit(q, c, name='gridScript')

    if (clicked_tile == TileItems.BOMB_UNEXPLODED):
        # hadamard gate applied to bomb qubit
        gridScript.h(q)

        # if there are more 1 hits then the bomb expodes and the game is lost
        if get_one_or_zero(gridScript, q, c) == 1:
            return TileItems.BOMB_EXPLODED
        return TileItems.BOMB_DEFUSED

    else:
        return clicked_tile


qcatsweeper/gui.py
from enum import Enum
from functools import partial

import qcatsweeper.quantum_logic as ql
import math
import pyxel


class GameState(Enum):
    INTRO = 0
    HELP = 1
    PLAYING_REAL = 2
    PLAYING_SIMULATED = 3
    LOST = 4
    WON = 5


def is_within(x, y, pos):
    x1, y1, x2, y2 = pos
    return x >= x1 and x <= x2 and y >= y1 and y <= y2

class QuantumCatsweeperApp:
    def __init__(self, width=149, height=165, debugging=False):
        # Initialize game state
        self.game_state = GameState.INTRO
        self.debugging = debugging

        self._main_tileset = 0
        self._exploding_cat_asset = 1

        self._main_bg = 1
        self._playing_bg = 2
        self._losing_bg = 3

        self._width = width
        self._height = height

        self._grid_size = 12
        self._grid_start_x = 5
        self._grid_start_y = 22
        self._grid_draw_size = 12

        self.game_grid = []
        self.game_grid_evaled = {}  # What string to display
        self.elapsed_frames = 0
        self.clicked_tiles = {}
        self.flagged_tiles = {}
        self.reveal_groups = {}
        self.golden_cat_x = -1
        self.golden_cat_y = -1

        pyxel.init(self._width, self._height, caption='Quantum Catsweeper')
        pyxel.mouse(True)
        self._play_real_button_pos = self.pyxel_button_centered(
            'Play', 100)
        self._help_button_pos = self.pyxel_button_centered(
            'Help', 120)
        self._help_back_button_pos = self.pyxel_button_centered('Back', 135)
        self._playing_real_back_button = self.pyxel_button('Back', 5, 5)
        self._replay_button = self.pyxel_button_centered('Play Again', 80)
        self._won_playagain_button = self.pyxel_button_centered(
            'Play Again', 100)

        pyxel.image(self._main_tileset).load(0, 0, 'assets/main_tileset.png')
        pyxel.image(self._exploding_cat_asset).load(
            0, 0, 'assets/explo_cat_31x25.png')

        # Sound
        pyxel.sound(self._main_bg).set(
            'e2e2c2g1 g1g1c2e2 d2d2d2g2 g2g2rr'
            'c2c2a1e1 e1e1a1c2 b1b1b1e2 e2e2rr', 'p', '3',
            'vffn fnff vffs vfnn', 25)
        pyxel.sound(self._playing_bg).set(
            'f0c1f0c1 g0d1g0d1 c1g1c1g1 a0e1a0e1'
            'f0c1f0c1 f0c1f0c1 g0d1g0d1 g0d1g0d1', 't', '7', 'n', 25)
        pyxel.sound(self._losing_bg).set(
            'c1g1c1g1 c1g1c1g1 b0g1b0g1 b0g1b0g1'
            'a0e1a0e1 a0e1a0e1 g0d1g0d1 g0d1g0d1', 't', '7', 'n', 25)

        pyxel.play(self._main_bg, [0, 1], loop=False)

        pyxel.run(self.update, self.draw)

    def update(self):
        if pyxel.btnp(pyxel.KEY_Q):
            pyxel.quit()

        if self.game_state == GameState.INTRO:
            self.handle_intro_events()

        elif self.game_state == GameState.HELP:
            self.handle_help_events()

        elif self.game_state == GameState.PLAYING_REAL:
            self.handle_playing_events()

        elif self.game_state == GameState.LOST:
            self.handle_lostgame_events()

        elif self.game_state == GameState.WON:
            self.handle_wongame_events()

    def draw(self):
        self.clear_assets()

        if self.game_state == GameState.INTRO:
            self.draw_introscreen()

        elif self.game_state == GameState.HELP:
            self.draw_helpscreen()

        elif self.game_state == GameState.PLAYING_REAL:
            self.draw_playscreen()

        elif self.game_state == GameState.LOST:
            self.draw_lostscreen()

        elif self.game_state == GameState.WON:
            self.draw_winscreen()

    def clear_assets(self):
        pyxel.cls(self._main_tileset)

    #### Event handlers ####

    def handle_wongame_events(self):
        if pyxel.btnp(pyxel.MOUSE_LEFT_BUTTON):
            mouse_within = partial(is_within, pyxel.mouse_x, pyxel.mouse_y)

            if mouse_within(self._won_playagain_button):
                self.game_state = GameState.PLAYING_REAL
                self.reset_game()
                pyxel.stop(self._main_bg)
                pyxel.play(self._playing_bg, [2, 3], loop=True)

            if mouse_within(self._playing_real_back_button):
                self.game_state = GameState.INTRO

    def handle_lostgame_events(self):
        if pyxel.btnp(pyxel.MOUSE_LEFT_BUTTON):
            mouse_within = partial(is_within, pyxel.mouse_x, pyxel.mouse_y)

            if mouse_within(self._playing_real_back_button):
                self.game_state = GameState.INTRO
                pyxel.stop(self._losing_bg)
                pyxel.play(self._main_bg, [0, 1], loop=True)

            elif mouse_within(self._replay_button):
                self.game_state = GameState.PLAYING_REAL
                self.reset_game()
                pyxel.stop(self._losing_bg)
                pyxel.play(self._playing_bg, [2, 3], loop=True)

    def is_flagged(self, pos):
        return pos in self.flagged_tiles and self.flagged_tiles[pos]

    def handle_playing_events(self):
        if pyxel.btnp(pyxel.MOUSE_RIGHT_BUTTON):
            # If user is clicking
            if pyxel.mouse_x > self._grid_start_x and pyxel.mouse_y > self._grid_start_y:
                row, col = self.get_grid_row_col_from_xy(
                    pyxel.mouse_x, pyxel.mouse_y)

                if row < 0 or row >= self._grid_size or col < 0 or col >= self._grid_size:
                    return

                clicked_tile = self.game_grid[row][col]

                if (row, col) not in self.clicked_tiles:
                    if (row, col) not in self.flagged_tiles:
                        self.flagged_tiles[(row, col)] = True
                    else:
                        self.flagged_tiles[(row, col)] = not self.flagged_tiles[(row, col)]

        if pyxel.btnp(pyxel.MOUSE_LEFT_BUTTON):
            mouse_within = partial(is_within, pyxel.mouse_x, pyxel.mouse_y)

            if mouse_within(self._playing_real_back_button):
                self.game_state = GameState.INTRO
                pyxel.stop(self._playing_bg)
                pyxel.play(self._main_bg, [0, 1], loop=True)

            # If user is clicking
            if pyxel.mouse_x > self._grid_start_x and pyxel.mouse_y > self._grid_start_y:
                row, col = self.get_grid_row_col_from_xy(
                    pyxel.mouse_x, pyxel.mouse_y)

                if row < 0 or row >= self._grid_size or col < 0 or col >= self._grid_size:
                    return

                clicked_tile = self.game_grid[row][col]

                if ((row, col) not in self.clicked_tiles) and not self.is_flagged((row, col)):
                    self.clicked_tiles[(row, col)] = True

                    if clicked_tile is ql.TileItems.GOLDEN_CAT:
                        self.game_state = GameState.WON
                        pyxel.stop(self._playing_bg)
                        pyxel.play(self._main_bg, [0, 1], loop=True)

                    # Call quantum computer to see if we reveal of nah
                    reveal_state = ql.onclick(clicked_tile)

                    # Move golden cat away from item
                    if reveal_state is ql.TileItems.NEG_EVAL:
                        offset_x = -1 if self.golden_cat_x < col else 1
                        offset_x = 0 if self.golden_cat_x == col else offset_x

                        offset_y = -1 if self.golden_cat_y < row else 1
                        offset_y = 0 if self.golden_cat_y == row else offset_y

                        # Move cat if the destination position is not clicked
                        if not self.swap_golden_cat_with(self.golden_cat_x + offset_x, self.golden_cat_y):
                            self.swap_golden_cat_with(
                                self.golden_cat_x, self.golden_cat_y + offset_y)

                    if reveal_state is ql.TileItems.POS_EVAL or \
                            reveal_state is ql.TileItems.BOMB_DEFUSED \
                            or reveal_state.value in ql.TileItems.TILE_NUMBERS.value:
                        # TODO: Move golden cat towards item
                        offset_x = 1 if self.golden_cat_x < col else -1
                        offset_x = 0 if self.golden_cat_x == col else offset_x

                        offset_y = 1 if self.golden_cat_y < row else -1
                        offset_y = 0 if self.golden_cat_y == row else offset_y

                        # Move cat if the destination position is not clicked
                        if not self.swap_golden_cat_with(self.golden_cat_x + offset_x, self.golden_cat_y):
                            self.swap_golden_cat_with(
                                self.golden_cat_x, self.golden_cat_y + offset_y)

                    # When bomb doesn't explode it turns into blank
                    if reveal_state is ql.TileItems.BOMB_DEFUSED:
                        self.game_grid[row][col] = ql.TileItems.BOMB_DEFUSED

                    if reveal_state is ql.TileItems.BOMB_EXPLODED:
                        self.game_grid[row][col] = ql.TileItems.BOMB_EXPLODED
                        self.game_state = GameState.LOST
                        pyxel.stop(self._playing_bg)
                        pyxel.play(self._losing_bg, 4, loop=True)

    def handle_help_events(self):
        if pyxel.btnp(pyxel.MOUSE_LEFT_BUTTON):
            mouse_within = partial(is_within, pyxel.mouse_x, pyxel.mouse_y)

            if mouse_within(self._help_back_button_pos):
                self.game_state = GameState.INTRO

    def handle_intro_events(self):
        if pyxel.btnp(pyxel.MOUSE_LEFT_BUTTON):
            mouse_within = partial(is_within, pyxel.mouse_x, pyxel.mouse_y)


            if mouse_within(self._play_real_button_pos):
                self.reset_game()
                self.game_state = GameState.PLAYING_REAL

                pyxel.stop(self._main_bg)
                pyxel.play(self._playing_bg, [2, 3], loop=True)

            if mouse_within(self._help_button_pos):
                self.game_state = GameState.HELP

    #### Screen Drawing ####

    def draw_winscreen(self):
        self.draw_grid()
        self.pyxel_button('Back', 5, 5)
        self.pyxel_text_centered(8, '       CONGRATULATIONS YOU WON', 3)
        # self.pyxel_text_centered(0, 'YOU WON!', 3)
        self.pyxel_button_centered('Play Again', 100)

    def draw_lostscreen(self):
        self.draw_grid()
        self.pyxel_button('Back', 5, 5)
        self.pyxel_button_centered('Play Again', 80)
        pyxel.blt(60, 52, self._exploding_cat_asset, 0, 0, 31, 25, 8)
        pyxel.text(60, 8, 'GAMEOVER', 8)

    def draw_grid(self):
        for row in range(len(self.game_grid)):
            for col in range(len(self.game_grid[row])):
                _x, _y = self.get_grid_xy_from_row_col(col, row)

                cur_tile = self.game_grid[row][col]
                if self.clicked_tiles.get((row, col), -1) == True:
                    display_tile_text = "_empty"

                    if (row, col) in self.game_grid_evaled:
                        display_tile_text = self.game_grid_evaled[(row, col)]
                    else:
                        display_tile_text = str(
                            abs(self.game_grid[row][col].value))

                    if cur_tile is ql.TileItems.BLANKS:
                        pyxel.rect(_x, _y, _x + self._grid_draw_size -
                                   2, _y - 2 + self._grid_draw_size, 6)

                    elif cur_tile is ql.TileItems.TILE1:
                        pyxel.rect(_x, _y, _x + self._grid_draw_size -
                                   2, _y - 2 + self._grid_draw_size, 3)
                        pyxel.text(_x + 2, _y + 2, display_tile_text, 7)

                    elif cur_tile is ql.TileItems.TILE2:
                        pyxel.rect(_x, _y, _x + self._grid_draw_size -
                                   2, _y - 2 + self._grid_draw_size, 11)
                        pyxel.text(_x + 2, _y + 2, display_tile_text, 0)

                    elif cur_tile is ql.TileItems.TILE3:
                        pyxel.rect(_x, _y, _x + self._grid_draw_size -
                                   2, _y - 2 + self._grid_draw_size, 9)
                        pyxel.text(_x + 2, _y + 2, display_tile_text, 0)

                    elif cur_tile is ql.TileItems.TILE4:
                        pyxel.rect(_x, _y, _x + self._grid_draw_size -
                                   2, _y - 2 + self._grid_draw_size, 10)
                        pyxel.text(_x + 2, _y + 2, display_tile_text, 0)

                    elif cur_tile is ql.TileItems.TILE5:
                        pyxel.rect(_x, _y, _x + self._grid_draw_size -
                                   2, _y - 2 + self._grid_draw_size, 13)
                        pyxel.text(_x + 2, _y + 2, display_tile_text, 0)

                    elif cur_tile is ql.TileItems.TILE6:
                        pyxel.rect(_x, _y, _x + self._grid_draw_size -
                                   2, _y - 2 + self._grid_draw_size, 1)
                        pyxel.text(_x + 2, _y + 2, display_tile_text, 7)

                    elif cur_tile is ql.TileItems.TILE7:
                        pyxel.rect(_x, _y, _x + self._grid_draw_size -
                                   2, _y - 2 + self._grid_draw_size, 1)
                        pyxel.text(_x + 2, _y + 2, display_tile_text, 7)

                    elif cur_tile is ql.TileItems.TILE8:
                        pyxel.rect(_x, _y, _x + self._grid_draw_size -
                                   2, _y - 2 + self._grid_draw_size, 1)
                        pyxel.text(_x + 2, _y + 2, display_tile_text, 7)

                    # Golden Cat
                    elif cur_tile is ql.TileItems.GOLDEN_CAT:
                        pyxel.blt(_x, _y, self._main_tileset,
                                  16, 0, self._grid_draw_size - 1, self._grid_draw_size - 1, 11)

                    # Bomb defused
                    elif cur_tile is ql.TileItems.BOMB_DEFUSED:
                        pyxel.blt(_x, _y, self._main_tileset,
                                  0, 0, self._grid_draw_size - 1, self._grid_draw_size - 1, 11)

                    # Bomb exploded
                    elif cur_tile is ql.TileItems.BOMB_EXPLODED:
                        pyxel.blt(_x, _y, self._main_tileset,
                                  0, 0, self._grid_draw_size - 1, -1 * (self._grid_draw_size - 1), 11)

                # Golden Cat (debug)
                elif cur_tile is ql.TileItems.GOLDEN_CAT and self.debugging:
                        pyxel.rect(_x, _y, _x + self._grid_draw_size -
                                   2, _y - 2 + self._grid_draw_size, 1)
                        pyxel.text(_x + 2, _y + 2, 'G', 12)

                # Flags
                elif self.flagged_tiles.get((row, col), -1) == True:
                    pyxel.rect(_x, _y, _x + self._grid_draw_size -
                               2, _y - 2 + self._grid_draw_size, 5)
                    pyxel.blt(_x, _y, self._main_tileset,
                              32, 0, self._grid_draw_size - 1, self._grid_draw_size - 1, 11)


                # Unclicked
                else:
                    pyxel.rect(_x, _y, _x + self._grid_draw_size -
                               2, _y - 2 + self._grid_draw_size, 5)

    def draw_playscreen(self):
        self.draw_grid()

        # Top bar stuff
        self.pyxel_button('Back', 5, 5)
        pyxel.text(50, 8, 'CATSWEEPER 9000', 7)

        # Convert text to "00:00" format
        display_mins = int(self.elapsed_frames / (60 * 30))  # 30 FPS
        display_mins = '0' + \
            str(display_mins) if display_mins < 10 else str(display_mins)
        display_sec = int(self.elapsed_frames / 30 % 60)
        display_sec = '0' + \
            str(display_sec) if display_sec < 10 else str(display_sec)
        pyxel.text(125, 8, '{}:{}'.format(display_mins, display_sec), 8)

        self.elapsed_frames = self.elapsed_frames + 1

    def draw_helpscreen(self):
        self.pyxel_text_centered(20, 'HELP', pyxel.frame_count % 16)
        # Information
        self.pyxel_text_centered(40, 'Welcome to Catsweeper !', 7)
        self.pyxel_text_centered(50, 'Cat-bombs have 50% chance', 7)
        self.pyxel_text_centered(60, 'to explode. If you click', 7)
        self.pyxel_text_centered(70, 'on a cat-bomb and it is', 7)
        self.pyxel_text_centered(80, 'upside-down, you were lucky !', 7)
        self.pyxel_text_centered(90, 'Find the golden cat', 7)
        self.pyxel_text_centered(100, 'to win!', 7)
        self.pyxel_button_centered('Back', 135)

    def draw_introscreen(self):
        # Draw Intro Text
        self.pyxel_text_centered(42, 'QUANTUM', pyxel.frame_count % 16)
        self.pyxel_text_centered(52, 'CATSWEEPER', pyxel.frame_count % 16)

        # Draw Quantum Cat
        cat_x = math.floor(self._width / 2) - 8
        cat_offset = math.sin(pyxel.frame_count * 0.1) * 2
        cat_flip = 1 if cat_offset >= 0 else -1
        pyxel.blt(cat_x + cat_offset, 62, self._main_tileset,
                  0, 0, 16 * cat_flip, 16, 11)

        # Draw button
        self.pyxel_button_centered('Play', 100)
        self.pyxel_button_centered('Help', 120)

    #### Pyxel Function Wrappers ####

    def pyxel_text_centered(self, y, t, col=None):
        # Font is 4 pixels width
        offset = math.ceil(len(t) * 4 / 2)
        x = math.floor(self._width / 2) - offset
        pyxel.text(x, y, t, col)

    def pyxel_blt_centered(self, y, img, sx, sy, w, h, colrow):
        offset = math.ceil((w - sx) / 2)
        x = math.floor(self._width / 2) - offset
        pyxel.blt(x, y, img, sx, sy, w, h, colrow)

    def pyxel_button(self, text, x, y):
        x_offset = (len(text) * 4) + 6
        y_offset = 8
        # Doing this so I can just get position of my buttons easily
        try:
            pyxel.rect(x, y, x + x_offset, y + y_offset, 14)
            pyxel.text(x + 4, y + 2, text, 0)
        except AttributeError as e:
            pass
        return (x, y, x + x_offset, y + y_offset)

    def pyxel_button_centered(self, text, y):
        offset = math.ceil(len(text) * 4 / 2) + 3
        x = math.floor(self._width / 2) - offset
        return self.pyxel_button(text, x, y)

    def get_grid_xy_from_row_col(self, x, y):
        gx = self._grid_start_x + (self._grid_draw_size * x)
        gy = self._grid_start_y + (self._grid_draw_size * y)
        return gx, gy

    def get_grid_row_col_from_xy(self, x, y):
        col = (x - self._grid_start_x) / self._grid_draw_size
        row = (y - self._grid_start_y) / self._grid_draw_size
        return int(row), int(col)

    #### Game State ####
    def swap_golden_cat_with(self, x, y):
        if x == self.golden_cat_x and y == self.golden_cat_y:
            return False

        if x >= 0 and x < self._grid_size and y >= 0 and y < self._grid_size:
            if (y, x) not in self.clicked_tiles:
                if self.game_grid[y][x] not in self.reveal_groups:
                    _tmp = self.game_grid[y][x]
                    self.game_grid[y][x] = ql.TileItems.GOLDEN_CAT
                    self.game_grid[self.golden_cat_y][self.golden_cat_x] = _tmp

                    self.golden_cat_x = x
                    self.golden_cat_y = y
                    return True
        return False

    def reset_game(self):
        self.elapsed_frames = 0
        self.clicked_tiles = {}
        self.flagged_tiles = {}
        self.reveal_groups = {}
        self.game_grid_evaled = {}
        self.game_grid = ql.new_game_grid(self._grid_size, bomb_no=20)

        for r in range(self._grid_size):
            for c in range(self._grid_size):
                if self.game_grid[r][c] == ql.TileItems.GOLDEN_CAT:
                    self.golden_cat_x = c
                    self.golden_cat_y = r
                    break
